crates/bevy_ecs/macros/src/component.rs

```diff
@@ -66,12 +73,40 @@ pub fn derive_component(input: TokenStream) -> TokenStream {
         .predicates
         .push(parse_quote! { Self: Send + Sync + 'static });
 
+    let requires = &attrs.requires;
+    let mut register_required = Vec::with_capacity(attrs.requires.iter().len());
+    let mut register_recursive_requires = Vec::with_capacity(attrs.requires.iter().len());
+    if let Some(requires) = requires {
+        for require in requires {
+            let ident = &require.ident;
+            register_recursive_requires.push(quote! {
+                <#ident as Component>::register_required_components(components, storages, required_components);
+            });
+            if let Some(func) = &require.func {
+                register_required.push(quote! {
+                    required_components.register(components, storages, || { let x: #ident = #func().into(); x });
+                });
+            } else {
+                register_required.push(quote! {
+                    required_components.register(components, storages, <#ident as Default>::default);
+                });
+            }
+        }
+    }
```

Comment:
It seems this will silently succeeds if `requires` contains the same component multiple times. Wouldn't it be better if it raised an error/warning?

Reply:
This is _slightly_ nontrivial because using just the built-in hashset storage, by default theres no way to tell if a component was contributed by the current required component (redundant + likely invalid) or contributed by a previously visited component "higher" in the tree (totally valid).

We'd need to allocate and populate a second hashset during init (more runtime work), or store the component that contributed the required component in the current hashset, so it can be compared (less runtime work, but adds complexity to the registration API).

The latter strategy probably makes the most sense, provided we're willing to add that to the registration API.

Reply:
I suppose we could also just compare the identifier paths in the macro and fail statically. That would cut down on runtime overhead, at the cost of not catching cases where two paths point to the same identifier. I think that actually has my preference, given that this is per-bundle work being done. I prefer to minimize overhead as much as possible.

CONST TYPEID WHERE ARE YOU

Reply:
My comment was just about the identifier paths that the macro sees. Checking them is of course not perfect, but at least avoids clearly non-sensical attributes

Reply:
I've added this!